unicorn/include/cocaine/zookeeper/handler.hpp

```diff
@@ -0,0 +1,230 @@
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation; either version 2 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*/
+#ifndef ZOOKEEPER_HANDLER_HPP
+#define ZOOKEEPER_HANDLER_HPP
+
+#include "cocaine/zookeeper/zookeeper.hpp"
+#include "cocaine/zookeeper/exception.hpp"
+
+#include <cocaine/locked_ptr.hpp>
+
+#include <functional>
+#include <string>
+#include <memory>
+#include <vector>
+
+#include <unordered_map>
+#include <cassert>
+
+namespace zookeeper {
+
+typedef Stat node_stat;
+
+//Used to prevent direct creation of managed handlers derived classes.
+class handler_tag {
+private:
+    handler_tag() {}
+    friend class handler_scope_t;
+};
+
+class managed_handler_base_t {
+public:
+    // Tagged constructor.
+    // Used to disable implicit instantiation.
+    // All instantiations are made through scope object.
+    managed_handler_base_t(const handler_tag&) {}
+    managed_handler_base_t(const managed_handler_base_t& other) = delete;
+    managed_handler_base_t& operator=(const managed_handler_base_t& other) = delete;
+    virtual ~managed_handler_base_t() {}
+};
+
+class handler_dispatcher_t {
+public:
+    static
+    void
+        watcher_cb(zhandle_t* zh, int type, int state, const char* path, void* watcherCtx);
+
+    static
+    void
+        void_cb(int rc, const void *data);
+
+    static
+    void
+        stat_cb(int rc, const struct Stat *stat, const void *data);
+
+    static
+    void
+        data_cb(int rc, const char *value, int value_len, const struct Stat *stat, const void *data);
+
+    static
+    void
+        string_cb(int rc, const char *value, const void *data);
+
+    static
+    void
+        strings_stat_cb(int rc, const struct String_vector *strings, const struct Stat *stat, const void *data);
+
+private:
+    typedef std::shared_ptr<managed_handler_base_t> handler_ptr;
+
+    //Ugly hack. That is fixed in C++14 - we can search in set by convertible value. Now we use map.
+    typedef std::unordered_map<const managed_handler_base_t*, handler_ptr> storage_t;
+
+    friend
+    class handler_scope_t;
+
+    static handler_dispatcher_t& instance();
+
+    template<class T, class ...Args>
+    void
+    call(const void* callback, Args&& ...args) {
+        handler_ptr cb;
+        {
+            std::unique_lock<std::mutex> lock(storage_lock);
+            auto it = callbacks.find(reinterpret_cast<const managed_handler_base_t*>(callback));
+            if (it != callbacks.end()) {
+                cb = it->second;
+                assert(cb);
+            }
+        }
+        if(cb) {
+            std::dynamic_pointer_cast<T>(cb)->operator()(std::forward<Args>(args)...);
```

Comment:
This is impossible, but as we have multiple virtual inheritance we need to use dynamic_cast